server/src/main/java/org/opensearch/indices/replication/OngoingSegmentReplications.java

```diff
@@ -139,13 +139,23 @@ void startSegmentCopy(GetSegmentFilesRequest request, ActionListener<GetSegmentF
      */
     CopyState prepareForReplication(CheckpointInfoRequest request, FileChunkWriter fileChunkWriter) throws IOException {
         final CopyState copyState = getCachedCopyState(request.getCheckpoint());
-        allocationIdToHandlers.compute(request.getTargetAllocationId(), (allocationId, segrepHandler) -> {
-            if (segrepHandler != null) {
-                logger.warn("Override handler for allocation id {}", request.getTargetAllocationId());
-                cancelHandlers(handler -> handler.getAllocationId().equals(request.getTargetAllocationId()), "cancel due to retry");
-            }
-            return createTargetHandler(request.getTargetNode(), copyState, request.getTargetAllocationId(), fileChunkWriter);
```

Comment:
>This is incorrectly using Map.compute which will cancel the existing handler but not replace its entry in the allocationIdToHandlers map.

I see the change in this PR is doing the same logical steps as done by `compute` eval here. The new value returned from remapping function here is added back again onto map. 
`allocationIdToHandlers.put(request.allocationId(), newValue);`

Reply:
The issue here is we are removing the handler from the map and inside of the compute function.  Internally the compute function will be applied before replacing the old value with a new value with the output of the computed value.  The compute function will resolve to "value2" but not before removing the key from the map.  Internally the map will then try to replace the key with "value2" but the key does not exist in the map any longer.

from compute [docs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#compute-K-java.util.function.BiFunction-)
```java
The default implementation is equivalent to performing the following steps for this map, then returning the current value or null if absent:
 
 V oldValue = map.get(key);
 V newValue = remappingFunction.apply(key, oldValue);
 if (oldValue != null ) {
    if (newValue != null)
       map.replace(key, oldValue, newValue);
    else
       map.remove(key, oldValue);
 } else {
    if (newValue != null)
       map.putIfAbsent(key, newValue);
    else
       return null;
 }
``` 

With simple example:
```java
    public void test() {
        Map<String, String> map = new ConcurrentHashMap<>();
        map.put("key", "value1");
        map.compute("key", (k, v) -> {
            if (v != null) {
                map.remove(k);
            }
            return "value2";
        });
        logger.info("Map contents: {}", map.entrySet());
    }
```
`[test] Map contents: []`